bolt-sidecar/src/state/mod.rs

```diff
@@ -1,32 +1,68 @@
 //! The `state` module is responsible for keeping a local copy of relevant state that is needed
 //! to simulate commitments against. It is updated on every block. It has both execution state and consensus state.
 
-use alloy_transport::TransportError;
-use thiserror::Error;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{future::poll_fn, Future, FutureExt};
+use tokio::time::Sleep;
 
 mod execution;
 pub use execution::{ExecutionState, ValidationError};
 
 /// Module to fetch state from the Execution layer.
 pub mod fetcher;
+pub use fetcher::StateClient;
 
 pub mod consensus;
 pub use consensus::ConsensusState;
 
 /// Module to track the head of the chain.
 pub mod head_tracker;
+pub use head_tracker::HeadTracker;
 
-/// Errors that can occur in the state module.
-#[derive(Debug, Error)]
-pub enum StateError {
-    /// An error occurred while fetching from an RPC client.
-    #[error("RPC error: {0:?}")]
-    Rpc(#[from] TransportError),
+/// The deadline for a which a commitment is considered valid.
+#[derive(Debug)]
+pub struct CommitmentDeadline {
+    slot: u64,
+    sleep: Option<Pin<Box<Sleep>>>,
 }
 
-#[derive(Debug, Clone)]
-struct ProposerDuties {
-    assigned_slots: Vec<u64>,
+impl CommitmentDeadline {
+    /// Create a new deadline for a given slot and duration.
+    pub fn new(slot: u64, duration: Duration) -> Self {
+        let sleep = Some(Box::pin(tokio::time::sleep(duration)));
+        Self { slot, sleep }
+    }
+
+    /// Poll the deadline until it is reached.
+    pub async fn wait(&mut self) -> Option<u64> {
+        let slot = poll_fn(|cx| self.poll_unpin(cx)).await;
+        self.sleep = None;
+        slot
+    }
```

Comment:
note to self: review the efficiency of this future, in the context of being polled inside a `tokio::select` branch

Reply:
Why not just await `self.sleep` directly?

Reply:
because self.sleep is an Option